[FLINK-12747][docs] Getting Started - Table API Example Walkthrough

sjwiesman opened the pull request:
Based on #8873 

## What is the purpose of the change

As part of Flip-42 we are introducing a new table based getting started guide. This introduction assumes no previous experience with Flink or stream processing and works within an IDE. 

## Brief changelog

ecbafb6  Adds a new module `flink-walkthrough-common` that users will include as a dependency when following along with the tutorial. It includes custom sources, sinks and UDFs which allows us to take advantage of more advanced features without exposing details to new users.

1182ddf Adds a table walkthrough archetype. Similar to the existing quickstart archetypes it allows users to quickly get a working project. It is different in that it includes extra dependencies and code specific to this walkthrough such as setting up the table environment and registering sources / sinks.

92c65d3 Adds the actual walkthrough under the getting started section of the docs.

sjwiesman commented on pom.xml:
```diff
@@ -82,11 +81,13 @@ under the License.
 		<module>flink-docs</module>
 		<module>flink-python</module>
 		<module>flink-ml-parent</module>
+		<module>flink-walkthroughs</module>
```
I don't know why the pom diff is so crazy, I think it's because this is based on #8873 which does not have the most recent pom. I've verified that this line is the only difference between this branch and master.

sjwiesman replied:
Rebased on fixed.

sjwiesman commented:
cc @knaufk @fhueske

sjwiesman commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+
+This defines our application as using one hour tumbling windows based on the timestamp column.
+So a row with timestamp `2019-06-01 01:23:47` will be put in the `2019-06-01 01:00:00` window.
+While semantically equivalent, when using windows:
```
This list of reasons for preferring windows is the section I am most unsure about. I'm am writing to an audience that has no idea what a window is and I'm not sure this clearly expresses the benefits.

knaufk replied:
Something like: Aggregations based on time are special, because time (as opposed to other arbitrary attributes) generally moves forward in a continuous streaming application. This enables Flink to perform certain optimizations such as state clean up, when using time window-based aggregations.

knaufk commented:
Thanks a lot @sjwiesman  for the contribution! A few general thoughts: 

* I really like that it's not the typical generic "wordcount"/"windowed wordcount"/etc.

* I would be in favor of simply adding this as an example to `flink-examples-table` without a maven archtype. Firstly, we already have quickstart modules to create program skeletons. Adding an additional option for the walkthroughs adds extra burden in terms of ongoing maintenance and might confuse users. Secondly, in my opinion a "code walkthrough" should focus on explainoing the code of an example so that I can easily follow it. I don't necessarily expect the utilities to bootstrap a new program here. (We should have a separate dedicated subsection in the "Getting Started" section for the quickstarts/project setup.)

* Personally I think, the example for a code walkthrough should be no-magic and I should be able to understand as much as possible as a new user. Here this means to go without the custom `TableSource`, which is not easy to understand and it is not clear why this can be a streaming and batch source at the same time etc.. Alternatively, one could simply read from a file of transactions (batch case) or read from a message queue (streaming case; maybe provide a minimal data generator). This would also go well with the future vision of "simply replace the source and your job turns into a streaming program", although we are not there yet.

* Do we also want to add SQL to this example (probably as a section in the end implementing the same business logic)? Currently, we don't plan to have dedicated "SQL Walkthrough". Do we need an SQL walkthrough at all?

What do you think? What do others think?

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+Output tables can be used to emit the result of a Table API or SQL query to an external system.
+Tables can support batch queries, streaming queries, or both. 
+
+{% highlight java %}
```
We should also add Scala here, shouldn't we?

knaufk replied:
The "Walkthroughs" somewhat replace the current "tutorials", so we should also have Python Table API as a third option here.

sjwiesman replied:
I agree but I would like to get some more feedback before investing the time here.

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+
+Next, we register tables that we can use for the input and output of our application. 
+The TableEnvironment maintains a catalog of tables that are registered by name. There are two types of tables, input tables and output tables.
+Input tables can be referenced in Table API and SQL queries and provide input data.
```
Maybe mention that this corresponds to "sinks" and "sources" in other frameworks/APIs.

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+    .insertInto("stdout");
+{% endhighlight %}
+
+This intial job simply reads all transactions and writes them to standard out. 
```
initial

knaufk replied:
output instead of "out"?

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+
+Now that we have the skeleton of a job, let's add some business logic.
+We want a report that shows the total spend for each account across each hour of the day.
+Just like a SQL query, we can select the required fields and group by our keys. 
```
an SQL query?

sjwiesman replied:
Its "a SQL query". The rule is that you use “a" before words that start with a consonant sound and “an” before words that begin with a vowel sound. English is weird.

knaufk replied:
I pronounced it incorrectly :) Like "as kwu al", not sequel.

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+The `timestamp` column represents the [event time]({{ site.baseurl }}/dev/event_time.html) of each row, where event time is the logical time when an event took place in the real world.
+Flink understands the concept of event time, which we can leverage to clean up our code. 
+
+Bucketing data based on time is very common operation in data processing, especially when working with infinite streams. 
```
Is "Bucketing" a more common term than simply "Grouping"?

sjwiesman replied:
They both make sense to me but I'm happy to change it.

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+Flink understands the concept of event time, which we can leverage to clean up our code. 
+
+Bucketing data based on time is very common operation in data processing, especially when working with infinite streams. 
+A bucket based on time is called a `Window` and Flink offers flexible windowing semantics. 
```
Technically, we also have non-time windows in Flink. Maybe simnply change to `TimeWindow`?

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+## Once More, With Streaming!
+
+With our buisness logic implemented it is time to go from batch to streaming.
+Because the Table Api supports running against both batch and streaming runners with the same semantics, migrating is as simple as changing the execution context.
```
API

knaufk replied:
I think, the story here should be "it is as simply as changing to a streaming source". The fact that the environment needs to change is a technical limitation at this point.

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,250 @@
+tEnv.registerTableSink("stdout", new StdOutTableSink());
+{% endhighlight %}
+
+We register two tables, a `transactions` input table and a `stdout` output table. 
```
The naming is a bit inconsistent: 
* `transactions` is named by its content
* `stdout` is named by the target system
We could also change `stdout` to something like `spend_reports`.

sjwiesman commented:
Thank you for taking a look @knaufk .

> I would be in favor of simply adding this as an example to flink-examples-table without a maven archtype. Firstly, we already have quickstart modules to create program skeletons. Adding an additional option for the walkthroughs adds extra burden in terms of ongoing maintenance and might confuse users. 

I believe the best walkthroughs are the ones users can easily run and make small modifications to. In practice, this means we need an easy way to import the initial project into their IDE. If the code is in `flink-examples-table` it is non-trivial to extricate it from the rest of flink. This archetype is not meant to replace the quickstart archetypes in any way, and only serves as a way to quickly and easily create a new project with the right dependencies and initial set up for following along with the tutorial. We can make the `Setting up a Maven Project` more clear about what we are doing and direct users towards the generic quickstarts if they want to build their own projects. 


> Secondly, in my opinion, a "code walkthrough" should focus on explaining the code of an example so that I can easily follow it. I don't necessarily expect the utilities to bootstrap a new program here. (We should have a separate dedicated subsection in the "Getting Started" section for the quickstarts/project setup.)

Again, not the intention but I take this to mean what I have is not clear enough. I would also support quickly moving [Application Development -> Project Build Setup](https://github.com/apache/flink/tree/master/docs/dev/projectsetup) to the getting started section with a nice introduction. This way it should be more clear to users that if they want to start a new project look here, and if they want a follow along look there. The fact that both these components use maven archetypes to facilitate their goals is just a coincidence. 

> Personally, I think, the example for a code walkthrough should be no-magic and I should be able to understand as much as possible as a new user. Here this means to go without the custom TableSource, which is not easy to understand and it is not clear why this can be a streaming and batch source at the same time etc.. Alternatively, one could simply read from a file of transactions (batch case) or read from a message queue (streaming case; maybe provide a minimal data generator). This would also go well with the future vision of "simply replace the source and your job turns into a streaming program", although we are not there yet.

1) This is not a dockerized environment and so I think we should avoid any external dependencies. 

2) I should better explain that tables are logical views on top of external systems such as file systems or message queues but I don't see them as magic. 

> Do we also want to add SQL to this example (probably as a section in the end implementing the same business logic)? Currently, we don't plan to have dedicated "SQL Walkthrough". Do we need an SQL walkthrough at all?

My thought was this would be just focused on the Table Api and we could do some sort of SQL walkthrough with the docker environment.

knaufk commented:
> My thought was this would be just focused on the Table Api and we could do some sort of SQL walkthrough with the docker environment.

Ok.

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+* This will be replaced by the TOC
+{:toc}
+
+## What Are We Building? 
```
It's a general good practice in technical writing to address the user as "you" and avoid using "we". Could the content be rephrased to accommodate this?

sjwiesman replied:
I agree in general, but I was making the change and it did sound right. I think the reason why is I don't view this a technical document so much as a conversation. We are taking the user through a tour of a sample application, skipping most of the details in favor of high-level concepts. Does that make sense or do I sound crazy?

morsapaes replied:
Thinking of just my preference, I would agree (so no — not crazy). Thinking about making documentation consistent in general (and considering this documentation first and a conversation second), I think it would be appropriate to cut the "we".

sjwiesman replied:
Ok, I think we're in agreement. It will take me a little bit to remove "we"'s in a way that makes sense. In the meantime, let me know if you have any other comments content-wise.

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+{% endhighlight %}
+
+We register two tables, a `transactions` input table and a `spend_report` output table. 
+The `transactions` table lets us read credit card transactions, which contain `accountId`'s, `timestamp`'s, and dollar `amount`'s. 
```
For readability, saying: "The transactions table (`transactions`) (...) account IDs (`accountID`), timestamps (`timestamp`) and US$ amounts (`amount`)"; could help.

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+
+#### Registering A UDF
+
+Along with tables we include a user defined function for working with timestamps. Our function takes a timestamp and rounds it down the nearest hour. 
```
Missing a "to" before "the nearest hour"?

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+The `timestamp` column represents the [event time]({{ site.baseurl }}/dev/event_time.html) of each row, where event time is the logical time when an event took place in the real world.
+Flink understands the concept of event time, which we can leverage to clean up our code. 
+
+Grouping data based on time is very common operation in data processing, especially when working with infinite streams. 
```
"a very"?

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+Flink understands the concept of event time, which we can leverage to clean up our code. 
+
+Grouping data based on time is very common operation in data processing, especially when working with infinite streams. 
+A grouping based on time is called a `Window` and Flink offers flexible windowing semantics. 
```
Could also use a support link.

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+
+## Once More, With Streaming!
+
+With our buisness logic implemented it is time to go from batch to streaming.
```
"business"?

morsapaes commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,243 @@
+StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+{% endhighlight %}
+
+And thats it, a fully functional, stateful, distributed streaming application!
```
'

sjwiesman commented:
I believe I have addressed all the comments including adding a scala version.

zentol commented:
CI report for commit b2821a6ae97fd943f3a66b672e85fbd2374126c4: FAILURE [Travis](https://travis-ci.org/flink-ci/flink-ci/builds/554872384?utm_source=github_status&utm_medium=notification)

knaufk commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/src/main/scala/SpendReport.scala:
```diff
@@ -0,0 +1,41 @@
+    val tEnv = BatchTableEnvironment.create(env)
+
+    tEnv.registerTableSource("transactions", new TransactionTableSource)
+    tEnv.registerTableSink("stdout", new SpendReportTableSink)
```
spend_report

knaufk commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+					<version>${scala.version}</version>
+					<scope>compile</scope>
+				</dependency>
+				<dependency>
```
This should not be needed.

NicoK replied:
this is still there

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+{% highlight scala %}
+tEnv
+    .scan("transactions")
+    .window(Tumble over 1.hour on 'timestamp as 'w)
```
add import for Tumble?

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+val env = StreamExecutionEnvironment.getExecutionEnvironment();
```
add imports? in particular for the scala examples, this is not completely straight forward.

knaufk replied:
`StreamExecutionEnvironment.getExecutionEnvironment;`

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+object SpendReport {
+
+    def main(args: Array[String]): Unit = {
+        val env = StreamExecutionEnvironment.getExecutionEnvironment();
```
StreamExecutionEnvironment.getExecutionEnvironment;

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+{% highlight scala %}
+package spendreport;
+
+import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
```
`import org.apache.flink.streaming.api.TimeCharacteristic`

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+{% highlight java %}
+package spendreport;
+
+import org.apache.flink.walkthrough.common.table.SpendReportTableSink;
```
`import org.apache.flink.streaming.api.TimeCharacteristic`

knaufk commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+title: "Table API"
+nav-id: tableapitutorials
+nav-title: 'Table API'
+nav-parent_id: apitutorials
```
Does it make sense to already put this under it's one section "Code Walkthroughs"?

sjwiesman commented:
@morsapaes I believe I've addressed all of your comments

@knaufk I'm working on adding testing to the archetypes, in particular the quickstarts have [archetype integration tests](https://maven.apache.org/archetype/maven-archetype-plugin/integration-test-mojo.html) but it doesn't seem like they actually run and that the quickstarts are covered by end to end tests. I am working to see if I can get the plugin to work since it is much lighter weight but worst case I will add the e2e tests.

sjwiesman commented on docs/getting-started/tutorials/table_api.md:
```diff
@@ -0,0 +1,423 @@
+{% highlight java %}
+tEnv
+    .scan("transactions")
+    .window(Tumble.over("1.hour").on("timestamp").as("w"))
```
I was thinking code snippets throughout with a full runnable project (with imports) at the end.

sjwiesman commented:
This thread has gotten a little unwieldy so I'd like to summarize where everything stands. 

There are three new modules: 
* `flink-walkthrough-common`: Pre-defined sources/sinks/UDFs that users will use when following along
* `flink-walkthrough-table-java`: A java maven archetype for initializing the walkthrough
* `flink-walkthrough-table-scala`: A scala maven archetype for initializing the walkthrough
* End-To-End tests that validate that both archetypes compile and run
* Walkthrough documentation that covers both java and scala using the table api
   * The python table api is not currently feature-complete enough to be included
   * The walkthrough contains code snippits throughout with the full runnable code at the end

NicoK commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+	<version>${version}</version>
+	<packaging>jar</packaging>
+
+	<name>Flink Walkthrough Table</name>
```
still needs to be addressed

NicoK commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+			<artifactId>scala-library</artifactId>
+			<version>${scala.version}</version>
+			<scope>provided</scope>
+		</dependency>
```
dependency on scala can be removed here (dragged in via transitive dependency)

NicoK commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+			<artifactId>flink-streaming-scala_${scala.binary.version}</artifactId>
+			<version>${flink.version}</version>
+			<scope>provided</scope>
+		</dependency>
```
`flink-streaming-scala` is also in here twice

NicoK commented:
eventually (probably at the end of this PR), you should also add the chinese placeholders for the new pages, since you were already adapting the navigation there

sjwiesman commented:
@NicoK Thank you for the review, my intention was to add table_api.zh.md at the very end so I wouldn't need to keep both files in sync.

sjwiesman commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,445 @@
+
+You can edit the `groupId`, `artifactId` and `package` if you like. With the above parameters,
+Maven will create a project with all the dependencies to complete this tutorial.
+After importing the project into your editor, you will see a file following code. 
```
yikes, good catch

sjwiesman commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,445 @@
+</div>
+</div>
+
+Two tables are registered, a transaction input table, and a spend report output table.
```
Actually this should be a semicolon ";" but good catch

sjwiesman commented on flink-walkthroughs/flink-walkthrough-common/src/main/java/org/apache/flink/walkthrough/common/table/SpendReportTableSink.java:
```diff
@@ -0,0 +1,94 @@
+		return this;
+	}
+
+	private static String format(Row row) {
```
Thank you, did not know how to do that!

NicoK replied:
FYI: IntelliJ provided that for me as one option of fixing this warning

sjwiesman commented on flink-walkthroughs/flink-walkthrough-table-java/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,263 @@
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-table-planner_${scala.binary.version}</artifactId>
+			<version>${flink.version}</version>
+			<scope>provided</scope>
```
You can either do that or place `flink-table-uber` in `/lib`. I think putting flink deps in `/lib` is cleaner but this is targetted at new users and this should be optimized towards ease of deployment. +1

NicoK replied:
actually, with https://issues.apache.org/jira/browse/FLINK-13399 (will end up in 1.9), this uber jar will even already be in `lib` - otherwise we would have needed at least a notice box.

sjwiesman replied:
I've added myself as a watcher on that ticket. I will follow and update this pr accordingly.

sjwiesman commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+			<artifactId>flink-table-planner_${scala.binary.version}</artifactId>
+			<version>${flink.version}</version>
+			<scope>provided</scope>
+		</dependency>
```
The other option is to copy `flink-table-uber` from `/opt` to '/lib' which I think is a cleaner solution. I will add a note to the walkthrough.

sjwiesman commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,333 @@
+			<artifactId>flink-table-api-java-bridge_${scala.binary.version}</artifactId>
+			<version>${flink.version}</version>
+			<scope>provided</scope>
+		</dependency>
```
Yes, actually `flink-walkthrough-common` should only depend on `flink-table-common` but there are some classes that haven't been ported yet.

sjwiesman commented on flink-walkthroughs/flink-walkthrough-common/pom.xml:
```diff
@@ -0,0 +1,71 @@
+			<artifactId>flink-table-planner_${scala.binary.version}</artifactId>
+			<version>${project.version}</version>
+			<scope>provided</scope>
+		</dependency>
```
Same as above

sjwiesman commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,445 @@
+This defines our application as using one hour tumbling windows based on the timestamp column.
+So a row with timestamp `2019-06-01 01:23:47` is put in the `2019-06-01 01:00:00` window.
+
+Aggregations based on time are unique because time, as opposed to other attributes, generally moves forward in a continuous streaming application.
```
I've added an additional sentence that I think clarifies this

sjwiesman commented:
@NicoK I believe I have addressed all of your comments

NicoK commented on flink-walkthroughs/flink-walkthrough-common/src/main/java/org/apache/flink/walkthrough/common/source/TransactionSource.java:
```diff
@@ -0,0 +1,60 @@
+		super(new RateLimitedIterator<>(TransactionIterator.unbounded()));
+	}
+
+	private static class RateLimitedIterator<T> implements Iterator<T>, Serializable {
```
seems this is still missing

NicoK commented on flink-walkthroughs/flink-walkthrough-common/src/main/java/org/apache/flink/walkthrough/common/sink/LoggerOutputFormat.java:
```diff
@@ -0,0 +1,47 @@
+ * A basic output format that logs all records at level <b>INFO</b>.
+ */
+@Internal
+public class LoggerOutputFormat implements OutputFormat<String> {
```
```suggestion
public class LoggerOutputFormat implements OutputFormat<String> {

	private static final long serialVersionUID = 1L;
```

NicoK replied:
still missing

sjwiesman commented:
Sorry about that @NicoK, should be taken care of now.

fhueske commented on flink-walkthroughs/flink-walkthrough-table-java/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,263 @@
+	<packaging>jar</packaging>
+
+	<name>Flink Walkthrough Table Java</name>
+	<url>http://www.myorganization.org</url>
```
drop or `https://flink.apache.org`?

fhueske commented on flink-walkthroughs/flink-walkthrough-table-java/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,263 @@
+
+		<!-- Example:
+
+        <dependency>
```
Indention?

fhueske commented on flink-walkthroughs/flink-walkthrough-table-scala/src/main/resources/archetype-resources/pom.xml:
```diff
@@ -0,0 +1,300 @@
+	<packaging>jar</packaging>
+
+	<name>Flink Walkthrough Table Scala</name>
+	<url>http://www.myorganization.org</url>
```
drop or `https://flink.apache.org`?

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+under the License.
+-->
+
+Apache Flink offers a Table API as a unified, relational API for batch and stream processing, i.e., queries are executed with the same semantics on unbounded, real-time streams or bounded, recorded streams and produce the same results.
```
Not sure how many stream processing newcomers can make the connection from "bounded, recorded stream" to "batch data". Since this is for ppl who just get in touch with Flink, we might want to make that more explicit.

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+
+Two tables are registered; a transaction input table, and a spend report output table.
+The transactions (`transactions`) table lets us read credit card transactions, which contain account ID's (`accountId`), timestamps (`timestamp`), and US$ amounts (`amount`).
+In practice, the `BoundedTransactionTableSource` may be backed by a filesystem, a database, or any other static source.
```
Mention that the data in this tutorial is generated?

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+
+Aggregations based on time are unique because time, as opposed to other attributes, generally moves forward in a continuous streaming application.
+In a batch context, windows offer a convenient API for grouping records by a timestamp attribute.
+When used in a streaming application, time window-based aggregations enables Flink to perform specific optimizations such as state clean up when the framework knows that no more records will arrive for a particular window.
```
Maybe move this sentence to the next section where streaming is used?

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+
+## Once More, With Streaming!
+
+Because Flink's Table API offers consistent semantics for both batch and streaming, migrating from one to the other requires just two steps.
```
`consistent semantics` -> `consistent syntax and semantics`?

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+</div>
+</div>
+
+And that's it, a fully functional, stateful, distributed streaming application!
```
Briefly explain what happens when the program is executed?

Maybe something like:
> The query continuously consumes the stream of transactions, computes the hourly spendings, and emits the result. Since the input is unbounded, the query keeps running until it is manually stopped.

Might also be good to quickly describe the results for each of the previous steps.

sjwiesman replied:
That makes sense, and then I can also work in the sentence about streaming windows here.

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+package spendreport;
```
rm `;`

fhueske commented on docs/getting-started/walkthroughs/table_api.md:
```diff
@@ -0,0 +1,444 @@
+        val env = StreamExecutionEnvironment.getExecutionEnvironment
+        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
+
+        val tEnv = StreamTableEnvironment.create(env);
```
rm `;`

sjwiesman commented:
Thanks for taking a look @fhueske, I believe I have addressed all of your points. In particular, I've added descriptions and sample output after each modification to the job.

NicoK commented on flink-end-to-end-tests/test-scripts/common.sh:
```diff
@@ -49,7 +49,7 @@ cd $TEST_ROOT
 source "${TEST_INFRA_DIR}/common_utils.sh"
 
-NODENAME=${NODENAME:-`hostname -f`}
+NODENAME="localhost" #${NODENAME:-`hostname -f`}
```
please revert

sjwiesman commented:
Sorry about that @NicoK, resolved.

sjwiesman commented:
Rebased and squashed.
